### PR TITLE
`PlutusLedgerApi.V1.withCurrencySymbol`

### PR DESCRIPTION
**Pull request by Unisay**
Closes https://github.com/IntersectMBO/plutus/issues/5992

### CODE REVIEW

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs [hunk 1]
```diff
@@ -8,26 +10,24 @@
...
+{-# OPTIONS_GHC -fexpose-all-unfoldings #-}
 {-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
 {-# OPTIONS_GHC -fno-spec-constr #-}
 {-# OPTIONS_GHC -fno-specialise #-}
-{-# OPTIONS_GHC -fexpose-all-unfoldings #-}
-{-# LANGUAGE BlockArguments       #-}
-{-# LANGUAGE DataKinds            #-}
-{-# LANGUAGE TypeFamilies         #-}
-{-# LANGUAGE UndecidableInstances #-}
--- We need -fexpose-all-unfoldings to compile the Marlowe validator
```
Comment (zliu41) [inline plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:22]: This comment should be preserved

**Review by zliu41 — COMMENTED** [review]
Could you link to the issue? I don't remember why this is needed. Also please add tests.

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs (hunk 1 above)
Comment (Unisay) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:22]: My logic was this:
In the Plutus documentation we have a list of GHC options that should be enabled for Plugin to work. Many Plinth modules use them. We should use the recommended options everywhere and there is nothing that makes this particular instance special enough to require and extra comment.

Comment (zliu41) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:22]: `-fexpose-all-unfoldings` is not an option that should be enabled everywhere, as explained in https://plutus.cardano.intersectmbo.org/docs/using-plutus-tx/extensions-flags-pragmas

Comment (Unisay) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:22]: Ok, I made a mistake assuming that this option should be used everywhere in Plinth by default.

Comment (Unisay) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:22]: Do you know why `{-# INLINEABLE #-}` is not enough for the Marlowe validator?

Comment (zliu41) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:22]: Missing unfolding for something, I think - we never identified the root cause.

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs [hunk 2]
```diff
@@ -393,7 +408,7 @@ lovelaceValueOf v = Lovelace (valueOf v adaSymbol adaToken)
 {-# INLINABLE assetClassValue #-}
 -- | A 'Value' containing the given amount of the asset class.
 assetClassValue :: AssetClass -> Integer -> Value
-assetClassValue (AssetClass (c, t)) i = singleton c t i
+assetClassValue (AssetClass (c, t)) = singleton c t
```
Comment (Unisay) [inline plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs:411]: Eta reduction fixes hlint warning

**Review by zliu41 — APPROVED** [review]

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Value.hs [hunk 3]
```diff
@@ -21,11 +21,15 @@
...
 {-# OPTIONS_GHC -fno-specialise #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
 {-# OPTIONS_GHC -fexpose-all-unfoldings #-}
--- We need -fexpose-all-unfoldings to compile the Marlowe validator
--- with GHC 9.6.2.
+-- We need -fexpose-all-unfoldings to compile the Marlowe validator with GHC 9.6.2.
 -- TODO. Look into this more closely: see https://github.com/IntersectMBO/plutus/issues/6172.
 
+{-# HLINT ignore "Redundant if" #-}
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Value.hs:30]: I run hlint locally and generally find it helpful. 

Having hlint warning in the codebase is a noise for me as it breaks my workflow when I need to "Jump to the next problem" so I'd really like to either address warnings or suppress them. 

Suppressing warnings in `hlint.yaml` works for me, even though its not as convenient as choosing a "Ignore the XXX" in vscode.

> which make the compilation slower.
Is this really the case?

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs [hunk 4]
```diff
@@ -8,21 +10,22 @@
...
 {-# OPTIONS_GHC -fexpose-all-unfoldings #-}
-{-# LANGUAGE BlockArguments       #-}
-{-# LANGUAGE DataKinds            #-}
-{-# LANGUAGE TypeFamilies         #-}
-{-# LANGUAGE UndecidableInstances #-}
 -- We need -fexpose-all-unfoldings to compile the Marlowe validator
--- with GHC 9.6.2.
 -- TODO. Look into this more closely: see https://github.com/IntersectMBO/plutus/issues/6172.
 
+{-# HLINT ignore "Redundant if" #-}
```
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Value.hs]: moved to `hlint.yaml`

File: plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Value.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Unisay) [reply plutus-ledger-api/src/PlutusLedgerApi/V1/Data/Value.hs]: moved to `hlint.yaml`

**Pull request merged**